hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -321,31 +320,39 @@ public void run(boolean cancelled) throws IOException {
         } else {
           Channel channel = channelRef.get();
           if (channel == null) {
-            try {
-              Channel newChannel = connect();
-              if (!channelRef.compareAndSet(null, newChannel)) {
-                newChannel.close();
+            connect().addListener(new ChannelFutureListener() {
+              @Override public void operationComplete(ChannelFuture channelFuture) {
+                Channel ch = channelFuture.channel();
+                if (channelFuture.isSuccess()) {
+                  if (!channelRef.compareAndSet(null, ch)) {
+                    ch.close();
+                    ch = channelRef.get();
+                  }
+                  writeAndFlushToChannel(call, ch);
```

Comment:
I would do this:

```
if (channelRef.compareAndSet(null, ch)) {
  writeAndFlushToChannel(call, ch);
} else {
  ch.close();
}
```

Reply:
In that case you swallow the Call although you have a usable connection (else branch). We have to write the Call to the existing channel.

Reply:
If there is a concurrent shutdown, `ch = channelRef.get()` might return `null`, so you call `writeAndFlushToChannel` with `null` channel. That should be avoided.

Reply:
That method begins with a null check for exactly the same reason.

Reply:
Definitely can benefit from a comment here :). I had to read this a couple of times.

Essentially, if we have a non-null channel already set on the reference, close the channel we just opened via from the Future, and then write our call to the channel which was already open.

Crazy question: is it possible that `ch` and `channelRef.get()` refer the same Channel object in this code?

Reply:
> That method begins with a null check for exactly the same reason.

Where? Are talking about line 327? Because that `null` check does not matter. We are talking about race conditions and `Atomic*` in java does not use any locking.

Reply:
I'm talking about `writeAndFlushToChannel()` method:
https://github.com/apache/hbase/pull/4125/files/adb74551de58c383a3123eb8fe2486f39aef02ef..73ec449dccc7519b8b27a5aae0bd222e4aeb42d1#diff-992c1a6e405b5e258665b955d3f15df00944e3694ad52e92d739b10cd8df3409R344

Reply:
Thanks @joshelser for chiming in, let me give a quick update. I'm talking to @meszibalu "offline" to speed things up a bit.
1. I think what you're saying is a non issue. If compareAndSet() fails then must be something *different* set to the AtomicReference. I'm sure @meszibalu can confirm that.
2. The issue we're talking is opening multiple channels concurrently. Balazs explained an example of 1000 concurrent Calls coming to the client and they're concurrently initiating `connect()`, because channelRef is null. The connection takes long and we'll end up having 1000 connections to the server and 1000 Calls compete to set its channel to the Ref. One wins, the others closes their connection.

Need to solve 2 things here:
1. Make sure that no Call is falling through the cracks. All of them must be put on the winning connection.
2. How can we avoid opening 999 connections that will be evantually closed.

Balazs:
```
AtomicBoolean connecting = new AtomicBoolean();
if (!connecting.compareAndSet(false, true)) {
  // connect
}
```

Andor: ReentrantLock